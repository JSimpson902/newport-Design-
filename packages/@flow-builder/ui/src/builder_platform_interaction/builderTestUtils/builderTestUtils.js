export * from './screenEditorTestUtils.js';
export * from './commonTestUtils';
export * from './deepFindMatchers';
export * from './useAdvancedOptionsTestUtils';
export * from './goldObjectMatchers';
export * from './domTestUtils';
export * from './complexTypeTestUtils';
