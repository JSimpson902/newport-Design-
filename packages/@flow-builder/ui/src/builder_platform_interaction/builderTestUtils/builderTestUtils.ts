export * from './combobox';
export * from './commonTestUtils';
export * from './complexTypeTestUtils';
export * from './deepFindMatchers';
export * from './domTestUtils';
export * from './events';
export * from './goldObjectMatchers';
export * from './manuallyAssignVariablesCheckboxTestUtils';
export * from './screenEditorTestUtils';
export * from './selectors';

// @sa11y testing appears to take some extra time (beyond the given 5000ms default jest timeout)
export const ACCESSIBILITY_TEST_TIMEOUT_IN_MS = 15000;

/**
 * Mocks a label in jest
 *
 * @param jest - The jest instance
 * @param labelName - The label name
 * @returns the mock label
 */
export function mockLabel(jest, labelName: string) {
    jest.mock(
        `@salesforce/label/${labelName}`,
        () => {
            return { default: `${labelName} mock value` };
        },
        { virtual: true }
    );

    return `${labelName} mock value`;
}
