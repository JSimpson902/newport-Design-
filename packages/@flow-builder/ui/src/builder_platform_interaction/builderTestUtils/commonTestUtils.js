const DEFAULT_MAX_TICKS = 10;

/**
 * Returns a promise that is fulfilled when condition is truthy and rejected if we exceed given max number of event loop ticks
 *
 * @param {Function} condition should return truthy value when condition is met
 * @param {number} maxTicks max number of event loop ticks before the condition becomes truthy
 * @returns {Promise} if fulfilled, value is the condition returned value
 */
export const until = (condition, maxTicks = DEFAULT_MAX_TICKS) => {
    return Promise.resolve().then(() => {
        try {
            const result = condition();
            if (result) {
                return Promise.resolve(result);
            }
            if (maxTicks === 0) {
                return Promise.reject(Error('Condition not met'));
            }
        } catch (error) {
            if (maxTicks === 0) {
                return Promise.reject(error);
            }
        }
        return until(condition, maxTicks - 1);
    });
};

/**
 * Returns a promise that is fulfilled when expectations are met
 * @param {Function} expectations should throw an exception if an expectation is not met
 * @param {number} maxTicks max number of event loop ticks before the condition becomes truthy
 * @returns {Promise} fulfilled when expectations are met
 */
export const untilNoFailure = (expectations, maxTicks = DEFAULT_MAX_TICKS) => {
    return until(() => {
        expectations();
        return true;
    }, maxTicks);
};

/**
 * Returns a promise that is fulfilled after the given number of event loop ticks
 * @param {number} maxTicks number of event loop ticks before the promise is fulfilled
 * @returns {Promise} fulfilled after the given number of event loop ticks
 */
export const ticks = (maxTicks = DEFAULT_MAX_TICKS) => {
    return until(() => false, maxTicks).catch(() => {});
};