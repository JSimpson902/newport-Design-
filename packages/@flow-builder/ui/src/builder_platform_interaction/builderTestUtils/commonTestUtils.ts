import { createElement } from 'lwc';
import { setDocumentBodyChildren } from './domTestUtils';

const DEFAULT_MAX_TICKS = 10;

/**
 * Returns a promise that is fulfilled when condition is truthy and rejected if we exceed given max number of event loop ticks
 *
 * @param {Function} condition should return truthy value when condition is met
 * @param {number} maxTicks max number of event loop ticks before the condition becomes truthy
 * @returns {Promise} if fulfilled, value is the condition returned value
 */
export const until = (condition, maxTicks = DEFAULT_MAX_TICKS) =>
    Promise.resolve().then(() => {
        try {
            const result = condition();
            if (result) {
                return Promise.resolve(result);
            }
            if (maxTicks === 0) {
                return Promise.reject(Error('Condition not met'));
            }
        } catch (error) {
            if (maxTicks === 0) {
                return Promise.reject(error);
            }
        }
        return until(condition, maxTicks - 1);
    });

/**
 * Returns a promise that is fulfilled when expectations are met
 *
 * @param {Function} expectations should throw an exception if an expectation is not met
 * @param {number} maxTicks max number of event loop ticks before the condition becomes truthy
 * @returns {Promise} fulfilled when expectations are met
 */
export const untilNoFailure = (expectations, maxTicks = DEFAULT_MAX_TICKS) =>
    until(() => {
        expectations();
        return true;
    }, maxTicks);

/**
 * Returns a promise that is fulfilled after the given number of event loop ticks
 *
 * @param {number} maxTicks number of event loop ticks before the promise is fulfilled
 * @returns {Promise} fulfilled after the given number of event loop ticks
 */
export const ticks = (maxTicks = DEFAULT_MAX_TICKS) => until(() => false, maxTicks).catch(() => {});

/**
 * Utility function that waits 1 tick (can be used to wait for rendering after a component update)
 */
export async function tick() {
    await ticks(1);
}

/**
 * Updates a lightning component's props and waits for it to render
 *
 * @param cmp - The component
 * @param props - The props
 */
export async function updateComponent(cmp, props) {
    Object.assign(cmp, props);
    await tick();
}

/**
 * Create a new promise with status properties.
 *
 * @param promise current promise
 * @returns new promise with status properties
 */
export const makeQuerablePromise = (promise) => {
    // Don't modify any promise that has been already modified.
    if (promise.isResolved) {
        return promise;
    }

    let isPending = true;
    let isRejected = false;
    let isFulfilled = false;

    const result = promise.then(
        (value) => {
            isFulfilled = true;
            isPending = false;
            return value;
        },
        (e) => {
            isRejected = true;
            isPending = false;
            throw e;
        }
    );

    result.isFulfilled = () => isFulfilled;
    result.isPending = () => isPending;
    result.isRejected = () => isRejected;
    return result;
};

/**
 * Formats a string to camelcase
 *
 * @param s - the string to format
 * @returns the formatted string
 */
const camelCase = (s: string) => s.replace(/-[a-zA-Z]/gi, (match) => match.charAt(1).toUpperCase());

/**
 * Creates a component for use in jest tests
 *
 * @param tagName - The component name (with namespace eg: builder_platform_interaction-breadcrumbs)
 * @param options - The component properties
 * @param optionsOverride - The overridden component properties
 * @returns promise with the lightning element for the component as returned value
 */
export const createComponent = async (tagName, options = {}, optionsOverride = {}) => {
    options = { ...options, ...optionsOverride };

    const [namespace] = tagName.split('-');
    const moduleName = camelCase(tagName.replace(`${namespace}-`, ''));

    const el = createElement(tagName, {
        is: require(`${namespace}/${moduleName}`).default
    });

    Object.assign(el, options);

    setDocumentBodyChildren(el);
    await tick();
    return el;
};

/*
 * @param callback Jest mocked callback
 * @returns the mocked callback event detail object value
 */
export const getDetailPassedToEvent = (callback: { mock: { calls: { detail: {} }[][] } }) =>
    callback.mock.calls[0][0].detail;
