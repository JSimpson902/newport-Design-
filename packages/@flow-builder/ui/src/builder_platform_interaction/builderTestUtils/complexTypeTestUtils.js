import { FLOW_DATA_TYPE } from 'builder_platform_interaction/dataTypeLib';

export const expectComplexTypeFieldDescription = field => {
    // need a dataType and apiName. isCollection and label optional
    expect(field.dataType).toBeDefined();
    expect(field.apiName).toBeDefined();
    if (
        field.dataType === FLOW_DATA_TYPE.SOBJECT.value ||
        field.dataType === FLOW_DATA_TYPE.APEX
    ) {
        expect(field.subtype).toBeDefined();
    }
};

export const expectFieldsAreComplexTypeFieldDescriptions = fields => {
    for (const fieldName in fields) {
        if (Object.prototype.hasOwnProperty.call(fields, fieldName)) {
            const field = fields[fieldName];
            expectComplexTypeFieldDescription(field);
        }
    }
};
