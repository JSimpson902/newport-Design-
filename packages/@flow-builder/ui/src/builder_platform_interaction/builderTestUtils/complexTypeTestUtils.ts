// @ts-nocheck
import { FLOW_DATA_TYPE } from 'builder_platform_interaction/dataTypeLib';

/**
 * A complex type field description has a dataType, an apiName, optionally a subtype, label and isCollection
 *
 * @param field
 */
export const expectComplexTypeFieldDescription = (field) => {
    // need a dataType and apiName. isCollection and label optional
    expect(field.dataType).toBeDefined();
    expect(field.apiName).toBeDefined();
    if (field.dataType === FLOW_DATA_TYPE.SOBJECT.value || field.dataType === FLOW_DATA_TYPE.APEX.value) {
        expect(field.subtype).toBeDefined();
        expect(field.subtype).not.toBeNull();
    }
};

export const expectFieldsAreComplexTypeFieldDescriptions = (fields) => {
    for (const fieldName in fields) {
        if (Object.prototype.hasOwnProperty.call(fields, fieldName)) {
            const field = fields[fieldName];
            expectComplexTypeFieldDescription(field);
        }
    }
};
