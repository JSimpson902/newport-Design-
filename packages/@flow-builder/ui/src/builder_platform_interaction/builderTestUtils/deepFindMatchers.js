/**
 * Deep find given element in object
 * @param {Object} object
 * @param {Object} element
 * @param {Function} [isEqual] the comparison function
 * @param {string} [path=''] current path
 * @return {string|undefined} the path to the element in object
 */
export const deepFind = (object, element, isEqual = (e1, e2) => e1 === e2, path = 'object') => {
    if (!isEqual(element, element)) {
        // we won't find anything
        return undefined;
    }
    if (isEqual(object, element)) {
        return path;
    }
    if (object == null) {
        return undefined;
    }
    const iterable = typeof object[Symbol.iterator] === 'function';
    if (iterable && typeof object !== 'string') {
        let i = 0;
        for (const value of object) {
            const found = deepFind(value, element, isEqual, path + `[${i++}]`);
            if (found !== undefined) {
                return found;
            }
        }
    } else if (typeof object === 'object') {
        for (const key in object) {
            if (Object.prototype.hasOwnProperty.call(object, key)) {
                let found = deepFind(key, element, isEqual, path + `.${key}`);
                if (found !== undefined) {
                    return found;
                }
                found = deepFind(object[key], element, isEqual, path + `.${key}`);
                if (found !== undefined) {
                    return found;
                }
            }
        }
    }
    return undefined;
};

/**
 * Deep find a common element between object1 and object2
 * @param {Object} object1
 * @param {Object} object2
 * @param {Function} [isEqual] the comparison function
 * @return {[string]|undefined} first element is path in object1, second element is path in object2
 */
export const deepFindCommonElement = (object1, object2, isEqual = (e1, e2) => e1 === e2, pathObject1 = 'object1', pathObject2 = 'object2') => {
    const path = deepFind(object1, object2, isEqual, pathObject1);
    if (path !== undefined) {
        return [path, pathObject2];
    }
    const iterable = typeof object2[Symbol.iterator] === 'function';
    if (iterable && typeof object2 !== 'string') {
        let i = 0;
        for (const value of object2) {
            const paths = deepFindCommonElement(object1, value, isEqual, pathObject1, pathObject2 + `[${i}]`);
            if (paths !== undefined) {
                return paths;
            }
            i++;
        }
    } else if (typeof object2 === 'object') {
        for (const key in object2) {
            if (Object.prototype.hasOwnProperty.call(object2, key)) {
                let paths = deepFindCommonElement(object1, key, isEqual, pathObject1, pathObject2 + `.${key}`);
                if (paths !== undefined) {
                    return paths;
                }
                paths = deepFindCommonElement(object1, object2[key], isEqual, pathObject1, pathObject2 + `.${key}`);
                if (paths !== undefined) {
                    return paths;
                }
            }
        }
    }
    return undefined;
};

const isImmutable = (element) => element == null || typeof element === 'string' || typeof element === 'number' || typeof element === 'boolean';

export const deepFindMatchers = {
    toHaveNoCommonMutableObjectWith(received, object2) {
        const paths = deepFindCommonElement(received, object2, (e1, e2) =>  e1 === e2 && !isImmutable(e1));
        if (!paths) {
            return {
                message: () => 'objects do not have a common mutable object',
                pass: true,
            };
        }
        return {
            message: () => `objects have a common mutable object : ${paths[0]}===${paths[1]}`,
            pass: false,
        };
    },
};