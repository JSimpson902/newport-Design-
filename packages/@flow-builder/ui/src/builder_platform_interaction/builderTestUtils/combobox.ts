import Combobox from 'builder_platform_interaction/combobox';
import { clickEvent, LIGHTNING_COMPONENTS_SELECTORS, removeEvent, ticks } from './builderTestUtils';

/**
 * Return the pill component corresponding to the given combobox
 *
 * @param {Combobox} combobox - combobox component
 * @returns {HTMLElement | null} lightning pill component
 */
export const getComboboxPill = (combobox: any): any | null => {
    const { shadowRoot } = combobox;
    if (!shadowRoot) {
        throw new Error('combobox shadowRoot must not be null');
    }
    return shadowRoot.querySelector(LIGHTNING_COMPONENTS_SELECTORS.LIGHTNING_PILL);
};

/**
 * Click on pill to switch to merge field notation combobox mode
 *
 * @param {Combobox} combobox - current combobox
 * @returns {Promise<void>}
 */
// eslint-disable-next-line @lwc/lwc/no-async-await
export const clickPill = async (combobox: Combobox): Promise<void> => {
    const pill = getComboboxPill(combobox);
    if (pill) {
        pill.dispatchEvent(clickEvent());
        await ticks(1);
    }
};

/**
 * Remove pill to switch to merge field notation combobox mode and reset combobox menu data
 *
 * @param {Combobox} combobox - current combobox
 * @returns {Promise<void>}
 */
// eslint-disable-next-line @lwc/lwc/no-async-await
export const removePill = async (combobox: Combobox): Promise<void> => {
    const pill = getComboboxPill(combobox);
    if (pill) {
        pill.dispatchEvent(removeEvent());
        await ticks(1);
    }
};
