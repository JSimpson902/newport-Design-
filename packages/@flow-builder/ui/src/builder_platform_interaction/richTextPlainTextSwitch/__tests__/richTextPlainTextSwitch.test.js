import { createElement } from 'lwc';
import RichTextPlainTextSwitch, { TEXT_MODES } from '../richTextPlainTextSwitch';
import { ticks } from 'builder_platform_interaction/builderTestUtils';

function createComponentForTest(props) {
    const el = createElement('builder_platform_interaction-rich-text-plain-text-switch', {
        is: RichTextPlainTextSwitch
    });
    Object.assign(el, props);
    document.body.appendChild(el);
    return el;
}

const SelectPlainTextModeEvent = new CustomEvent('select', {
    detail: { value: TEXT_MODES.plainText }
});

const SelectRichTextModeEvent = new CustomEvent('select', {
    detail: { value: TEXT_MODES.richText }
});

const SELECTORS = {
    lightningButtonMenu: 'lightning-button-menu',
    lightningMenuItem: 'lightning-menu-item'
};

const getLightningButtonMenu = richTextPlainTextSwitchComponent => {
    return richTextPlainTextSwitchComponent.shadowRoot.querySelector(SELECTORS.lightningButtonMenu);
};
describe('plain-text-mode-select', () => {
    let richTextPlainTextSwitchComponent;
    beforeEach(() => {
        richTextPlainTextSwitchComponent = createComponentForTest();
    });
    describe('By default "RichText" mode selected', () => {
        test('check UI (snapshot) - menu items details: label, checked status', () => {
            expect(richTextPlainTextSwitchComponent).toMatchSnapshot();
        });
    });
    describe('Select "PlainText" mode', () => {
        it('select "plain text" mode via API', async () => {
            richTextPlainTextSwitchComponent.selectedMode = TEXT_MODES.plainText;
            await ticks(1);
            expect(richTextPlainTextSwitchComponent).toMatchSnapshot();
        });
        it('select "plain text" mode via UI', async () => {
            const lightningButtonMenu = getLightningButtonMenu(richTextPlainTextSwitchComponent);
            lightningButtonMenu.dispatchEvent(SelectPlainTextModeEvent);
            await ticks(1);
            expect(richTextPlainTextSwitchComponent).toMatchSnapshot();
        });
    });
    describe('Select "RichText" mode', () => {
        it('select "rich text" mode via API', async () => {
            richTextPlainTextSwitchComponent.selectedMode = TEXT_MODES.richText;
            await ticks(1);
            expect(richTextPlainTextSwitchComponent).toMatchSnapshot();
        });
        it('select "rich text" mode via UI', async () => {
            const lightningButtonMenu = getLightningButtonMenu(richTextPlainTextSwitchComponent);
            lightningButtonMenu.dispatchEvent(SelectRichTextModeEvent);
            await ticks(1);
            expect(richTextPlainTextSwitchComponent).toMatchSnapshot();
        });
    });
});
