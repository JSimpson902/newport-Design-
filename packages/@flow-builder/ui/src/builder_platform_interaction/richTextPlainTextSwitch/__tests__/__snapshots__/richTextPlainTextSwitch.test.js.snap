// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plain-text-mode-select By default "RichText" mode selected check UI (snapshot) - menu items details: label, checked status 1`] = `
<builder_platform_interaction-rich-text-plain-text-switch>
  #shadow-root(open)
    <div
      class="divButtonMenuPlainRichText slds-grid slds-grid_vertical-align-center"
    >
      <span
        class="spanSelectedModeLabel"
      >
        FlowBuilderResourcedRichTextEditor.richText
      </span>
      <lightning-button-menu>
        #shadow-root(open)
          <slot />
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
      </lightning-button-menu>
    </div>
</builder_platform_interaction-rich-text-plain-text-switch>
`;

exports[`plain-text-mode-select Select "PlainText" mode select "plain text" mode via API 1`] = `
<builder_platform_interaction-rich-text-plain-text-switch>
  #shadow-root(open)
    <div
      class="divButtonMenuPlainRichText slds-grid slds-grid_vertical-align-center"
    >
      <span
        class="spanSelectedModeLabel"
      >
        FlowBuilderResourcedRichTextEditor.plainText
      </span>
      <lightning-button-menu>
        #shadow-root(open)
          <slot />
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
      </lightning-button-menu>
    </div>
</builder_platform_interaction-rich-text-plain-text-switch>
`;

exports[`plain-text-mode-select Select "PlainText" mode select "plain text" mode via UI 1`] = `
<builder_platform_interaction-rich-text-plain-text-switch>
  #shadow-root(open)
    <div
      class="divButtonMenuPlainRichText slds-grid slds-grid_vertical-align-center"
    >
      <span
        class="spanSelectedModeLabel"
      >
        FlowBuilderResourcedRichTextEditor.plainText
      </span>
      <lightning-button-menu>
        #shadow-root(open)
          <slot />
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
      </lightning-button-menu>
    </div>
</builder_platform_interaction-rich-text-plain-text-switch>
`;

exports[`plain-text-mode-select Select "RichText" mode select "rich text" mode via API 1`] = `
<builder_platform_interaction-rich-text-plain-text-switch>
  #shadow-root(open)
    <div
      class="divButtonMenuPlainRichText slds-grid slds-grid_vertical-align-center"
    >
      <span
        class="spanSelectedModeLabel"
      >
        FlowBuilderResourcedRichTextEditor.richText
      </span>
      <lightning-button-menu>
        #shadow-root(open)
          <slot />
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
      </lightning-button-menu>
    </div>
</builder_platform_interaction-rich-text-plain-text-switch>
`;

exports[`plain-text-mode-select Select "RichText" mode select "rich text" mode via UI 1`] = `
<builder_platform_interaction-rich-text-plain-text-switch>
  #shadow-root(open)
    <div
      class="divButtonMenuPlainRichText slds-grid slds-grid_vertical-align-center"
    >
      <span
        class="spanSelectedModeLabel"
      >
        FlowBuilderResourcedRichTextEditor.richText
      </span>
      <lightning-button-menu>
        #shadow-root(open)
          <slot />
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
        <lightning-menu-item>
          #shadow-root(open)
        </lightning-menu-item>
      </lightning-button-menu>
    </div>
</builder_platform_interaction-rich-text-plain-text-switch>
`;
